### PR TITLE
EHN: Convection Contours and HMB

### PR DESCRIPTION
**Pull request by carleyjmartin**
# Scope 

This PR adds the HMB and contours to the convection maps. There are a number of discussion points for how we want to style this plot which I suggest we talk about at the next pyDARN meeting.

- [x] Contour locator: At the moment I have a FixedLocator for ever 5 kV from -100 to 100 with a gap (-1,1) to ensure 0 is white/transparent, works for my needs but probably not good for pyDARN.
- [x] Assess colormap needs, using a generic matplotlib green-white-purple for now, probably will require custom one. RST uses the same color for + and -, do we want this too? Also including colormap options in key words for user to use their own.
- [x] Edge color of contour plotting, I'm using contourf to plot the contours, but there's a bug when you use alpha<1 where the edged of the contours do not take alpha, there's a whole big discussion about it on matplotlibs github, at the moment I don't know how to fix this and don't think we want to go with alpha=1.
- [x] Not using filled contours at all, just contour edges with lables?
- [x] z-ordering for all components
- [x] Default options (FOV on/off, radar location on/off etc. etc.)

More information can be found on the related issue.

**issue:** #235 

## Approval

**Number of approvals:** 2+ (code review, science review and testing) Please test on southern data, I have no southern maps on hand. 

## Test

**matplotlib version**: 3.5.1
**Note testers: please indicate what version of matplotlib you are using**

```python
map_file = "/Users/carley/Documents/data/maps/20150310.n.map"
map_data = pydarn.SuperDARNRead().read_dmap(map_file)
pydarn.Maps.plot_mapdata(map_data,record=180, parameter=pydarn.MapParams.FITTED_VELOCITY, lowlat=60)
plt.show()
```
To return:
image

*Reminder, please check your code is:
  - copyrighted (if applicable)
  - Appropriate disclaimer added
  - Modification line filled in (if applicable)*

### CODE REVIEW

**Comment by carleyjmartin** [comment]
Thanks Emma! The colormap is just a place holder I chose at random, I don't want to be the person to make that executive decision haha, and definitely we should chat about it at the next meeting! Dan has also suggested that we only use contour lines rather than fill too, so lots to talk about next meeting I think.

File: pydarn/plotting/maps.py [hunk 1]
```diff
@@ -479,3 +501,206 @@ def index_legendre(l: int, m: int):
...
+            hemi = -1
+
+        # Make a grid of the space the potential is evaluated on
+        # in magnetic coordinates
+        lat_step = 1
+        lon_step = 2
+        num_lats = int((90.0 - lowlat) / lat_step) + 1
+        num_lons = int(360.0 / lon_step) + 1
+        lat_arr = np.array(range(num_lats)) * lat_step + lowlat
+        lat_arr = lat_arr * hemi
```
Comment (mts299) [inline pydarn/plotting/maps.py:587]:
``` suggestion
        lat_arr = lat_arr * hemisphere
```

File: pydarn/plotting/maps.py [hunk 2]
```diff
@@ -479,3 +501,206 @@ def index_legendre(l: int, m: int):
...
+
+        Parameters
+        ----------
+            ax: object
+                matplotlib axis object
+            mlats: List[float]
+                Magnetic Latitude in degrees
+            mlons: List[float]
+                Magnetic Longitude in radians
+
```
Comment (mts299) [inline pydarn/plotting/maps.py:522]: missing line colo and maybe linewidth is we set it to 1?

File: pydarn/plotting/maps.py [hunk 3]
```diff
@@ -479,3 +501,206 @@ def index_legendre(l: int, m: int):
...
+        """
+        plt.plot(mlons,mlats, c=line_color, linewidth=1)
```
Comment (mts299) [inline pydarn/plotting/maps.py:524]:
``` suggestion
        plt.plot(mlons,mlats, c=line_color, **kwargs)
```
linewidth can be passed in as a kwargs option so the user can change it.

**Review by mts299 — COMMENTED** [review]
Please run PEP8 (flask8) on it some stylistic issues going on here. If you cannot run it let me know and I will fix it up. 

Some other code clean up suggestions to make the code a bit cleaner and more flexibility. Otherwise looks good.

File: pydarn/plotting/maps.py [hunk 4]
```diff
@@ -479,3 +501,206 @@ def index_legendre(l: int, m: int):
...
+        except KeyError:
+            hemisphere = Hemisphere.North
+
+        # Lowest latitude to calculate potential to
+        theta_max = np.radians(90-np.abs(lat_min))
+
+        if hemisphere == Hemisphere.North:
+            hemi = 1
+        else:
+            hemi = -1
```
[the remark this answers is not in the document]
Comment (carleyjmartin) [reply pydarn/plotting/maps.py:578]: I thought it did but when I put the below suggestion in I get the error:
`TypeError: unsupported operand type(s) for *: 'int' and 'Hemisphere'`
So I'm not 100% on how to do that?

File: pydarn/plotting/maps.py [hunk 5]
```diff
@@ -479,3 +501,206 @@ def index_legendre(l: int, m: int):
...
+        # Get max value of potential
+        ind_max = np.where(pot_arr == pot_arr.max())
+        ind_min = np.where(pot_arr == pot_arr.min())
+        max_mlon = mlon[ind_max]
+        max_mlat = mlat[ind_max]
+        min_mlon = mlon[ind_min]
+        min_mlat = mlat[ind_min]
+
+        plt.scatter(np.radians(max_mlon),max_mlat,marker='+',s=70, color='k')
+        plt.scatter(np.radians(min_mlon),min_mlat,marker='x',s=70, color='k')
```
[the remark this answers is not in the document]
Comment (mts299) [reply pydarn/plotting/maps.py]: O crap you are right, forgot how dumb matplotlib is *shakes fist* 

NVM... just make it a parameter for the user to change I guess.... sigh

File: pydarn/plotting/maps.py (hunk 1 above)
Comment (carleyjmartin) [reply pydarn/plotting/maps.py]: Answer for future Carley who will come searching for this: `hemisphere.value`!

**Comment by carleyjmartin** [comment]
> @carleyjmartin would it be too much before the meeting to see if you can come up with a few other colour maps or designs on the contours?

I can try a few inbuilt ones, and show an example of a contour only maybe, I'll just post them here so people can see on a shared screen!

**Comment by carleyjmartin** [comment]
Here's some of the basic options, I can talk us through them in the meeting, I've not got time to make my own cmap but I can describe them.
Screen Shot 2022-03-17 at 9 05 58 AM
Screen Shot 2022-03-17 at 9 07 11 AM
38ef.png)
Screen Shot 2022-03-17 at 9 23 59 AM
Screen Shot 2022-03-17 at 9 26 03 AM

**Comment by carleyjmartin** [comment]
So I've worked on a few of the defaults we talked about today. 
Here is the default default:
Screen Shot 2022-03-17 at 4 46 22 PM
You'll notice there are tiny gaps in the contours... I'm trying to add the contour labels, and as you can see it removes a section for the labels, and I can even get the function to print the labels that should be there, but alas, the labels don't appear on the plot, so if anyone wants to try this out and see if they can get them, I'm at a loss because there's no reason they shouldn't be coming up. 

If you choose to set `contour_fill = True` you will get this as default:
Screen Shot 2022-03-17 at 4 49 28 PM
I'm going to add an optional colorbar for a chosen colormap too, the user can also put whatever cmap they want in.

There's a number of optional keywords I've added to allow some user input to get what they want, you can check that out at the bottom of maps.py.

I also need to look at the zvalues for the components still.

**Comment by mts299** [comment]
@ecbland  @Shirling-VT @KhanalKrishna @kherli Thoughts on the defaults and want to test it out?

**Comment by mts299** [comment]
@carleyjmartin do we want to make boundary false and radar labels false by default for convection maps?

**Comment by ecbland** [comment]
Looking good! (again!) 

It looks like the velocity colour bar label doesn't appear by default. Could that be added?

Figure_1

```
pydarn.Maps.plot_mapdata(map_data, record=400, boundary=False, radar_location=False, contour_fill=True, contour_colorbar=True,lowlat=50)
```

**Comment by ecbland** [comment]
I'm also getting some arrays printed in the terminal when I run `pydarn.Maps.plot_mapdata(...)`

```
(array([60.5, 59.5, 60.5, 61.5, 60.5, 61.5, 60.5, 61.5, 64.5, 60.5, 61.5,
       63.5, 64.5, 65.5, 61.5, 64.5, 65.5, 60.5, 63.5, 64.5, 65.5, 66.5,
       61.5, 63.5, 64.5, 60.5, 63.5, 65.5, 71.5, 72.5, 79.5, 79.5, 78.5,
       80.5, 80.5, 81.5, 81.5, 81.5, 75.5, 82.5, 82.5, 82.5, 76.5, 82.5,
       77.5, 83.5, 83.5, 83.5, 84.5, 84.5, 84.5, 85.5, 62.5, 63.5, 64.5,
       64.5, 65.5, 65.5, 66.5, 67.5, 67.5, 67.5, 68.5, 68.5, 69.5, 69.5,
       69.5, 70.5, 70.5, 70.5, 71.5, 71.5, 72.5, 72.5, 73.5, 73.5, 74.5,
       74.5, 75.5, 76.5, 75.5, 74.5, 75.5, 75.5, 75.5, 76.5, 76.5, 77.5,
       76.5, 77.5, 76.5, 77.5, 78.5, 75.5, 76.5, 77.5, 78.5, 79.5, 75.5,
...
```

Is that intentional?

**Comment by carleyjmartin** [comment]
@mts299 is the boundary keyword working with all the changes from develop merged in? I can check this this afternoon. I can't get the radar labels to work, so I can set it to false but if someone sets it to True they just make the gap for the labels but don't plot them there, should we just remove this functionality until we solve the problem?
*Should we have a default cmap color fill like the one example you have if the user just sets true for the fill but no cmap *- If someone sets contour_fill to True it should default to the blue/red cmap so the user doesn't have to enter a cmap. Is that what you mean?

@ecbland, sure thing! It's in a chunk of code that I didn't touch for this PR, but I don't think there will be any issue with me just adding it in there. 
I think the printed array might actually be a hold over from a stray print statement that got merged into the new develop recently, but it's probably easier to address here than opening a new PR to remove a single print statement!

Thanks both!

**Comment by carleyjmartin** [comment]
Hey all, just a quick reply on the contour labelling. 
It looks like I'm having issues with the clabel function in matplotlib because the function that locates the position on the contours to remove a section to add a label works in polar plots, but the function that finds the locations for the bbox to add the text doesn't seem to understand polar plots (e.g. the example I'm looking at is trying to position a text label at the position x = -2.6878070480712672, y =-93.82328779989909, which you can see if we're plotting in radians(lon) and lat is pretty much meaningless)

Anyway, the bad news here is I think for now matplotlib is broken for contour labels in this circumstance. 
I think in the future we could possibly create our own labelling system for the contours, it'll be a job but it's probably better for our purpose than the default way contour labels are done in matplotlib anyway.

**Pull request merged**